modules/javafx.graphics/src/main/java/com/sun/javafx/css/BinarySerializer.java

```diff
@@ -0,0 +1,186 @@
+        int type = is.readByte();
+
+        if (type == TYPE_SIMPLE) {
+            return readSimpleSelector(is, strings);
+        }
+
+        // Backwards compatible, if it is not TYPE_SIMPLE any other value is considered TYPE_COMPOUND
+        return readCompoundSelector(is, strings);
+    }
+
+    public static void write(Selector selector, DataOutputStream os, StyleConverter.StringStore stringStore) throws IOException {
+        if (selector instanceof SimpleSelector s) {
+            writeSimpleSelector(s, os, stringStore);
+        }
+        else if(selector instanceof CompoundSelector s) {
+            writeCompoundSelector(s, os, stringStore);
+        }
+        else {
+            throw new IllegalStateException("support missing for selector type: " + (selector == null ? "null" : selector.getClass()));
+        }
+    }
+
+    private static SimpleSelector readSimpleSelector(DataInputStream is, String[] strings) throws IOException {
+        String name = strings[is.readShort()];
+        int nStyleClasses = is.readShort();
+        List<String> styleClasses = new ArrayList<>();
+
+        for (int n = 0; n < nStyleClasses; n++) {
+            styleClasses.add(strings[is.readShort()]);
+        }
+
+        String id = strings[is.readShort()];
+        int nPseudoclasses = is.readShort();
+        List<String> pseudoclasses = new ArrayList<>();
+
+        for (int n = 0; n < nPseudoclasses; n++) {
+            pseudoclasses.add(strings[is.readShort()]);
+        }
+
+        return new SimpleSelector(name, styleClasses, pseudoclasses, id);
+    }
+
+    private static CompoundSelector readCompoundSelector(DataInputStream is, String[] strings) throws IOException {
+        int nSelectors = is.readShort();
+        List<SimpleSelector> selectors = new ArrayList<>();
+
+        for (int n = 0; n < nSelectors; n++) {
+
+            /*
+             * An extra byte is part of the binary format containing TYPE_SIMPLE that
+             * isn't strictly needed. However, to remain backwards compatible with the
+             * first (and only) supported binary format, this byte must be skipped over
+             * before attempting to read what is known to be a simple selector.
+             */
+
+            is.readByte();
+
+            selectors.add(readSimpleSelector(is, strings));
+        }
+
+        int nRelationships = is.readShort();
+
+        List<Combinator> relationships = new ArrayList<>();
+
+        for (int n = 0; n < nRelationships; n++) {
+            int ordinal = is.readByte();
+
+            if (ordinal == Combinator.CHILD.ordinal()) {
+                relationships.add(Combinator.CHILD);
+            }
+            else if (ordinal == Combinator.DESCENDANT.ordinal()) {
+                relationships.add(Combinator.DESCENDANT);
+            }
+            else {
+                assert false : "error deserializing CompoundSelector: Combinator = " + ordinal;
+                relationships.add(Combinator.DESCENDANT);
```

Comment:
throw an exception instead?

Reply:
This is original code, just moved.  Do you want me to change it still?